### PR TITLE
DCC-4636. Separate release and submission reports

### PR DESCRIPTION
**Pull request by slovit**

### CODE REVIEW

File: dcc-submission-core/src/main/java/org/icgc/dcc/submission/release/model/ReleaseView.java [hunk 1]
```diff
@@ -63,34 +57,10 @@ public ReleaseView(Release release, List<Project> projects, Map<String, List<Sub
...
-      detailedSubmission.setSubmissionFiles(submissionFiles.get(projectKey));
-      this.submissions.add(detailedSubmission);
-
-      Integer stateCount = this.summary.get(detailedSubmission.getState());
-      if (stateCount == null) {
-        stateCount = 0;
-      }
-      stateCount++;
-      this.summary.put(detailedSubmission.getState(), stateCount);
+    for (val submissionState : submissionsStates) {
```
Comment (btiernay) [inline dcc-submission-core/src/main/java/org/icgc/dcc/submission/release/model/ReleaseView.java]: http://stackoverflow.com/questions/25441088/group-by-counting-in-java8-stream-api:

``` java
val summary  = submissionsStates.stream().collect(groupingBy(identity(), counting()));
```

File: dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/repository/SubmissionRepository.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+  public List<Submission> findSubmissionsByProjectKey(@NonNull String projectKey) {
+    return query()
+        .where(entity.projectKey.eq(projectKey))
+        .list();
+  }
+
+  public int deleteByReleaseAndNotState(@NonNull String releaseName, @NonNull SubmissionState state) {
+    val result = delete(createQuery()
+        .filter("releaseName == ", releaseName)
```
Comment (btiernay) [inline dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/repository/SubmissionRepository.java:123]: Can we use type safe morphia dsl?

File: dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/service/ReleaseService.java [hunk 3]
```diff
@@ -244,7 +204,15 @@ public void createInitialRelease(Release initRelease) {
...
-    submissionFileSystem.createInitialReleaseFilesystem(nextRelease, projects);
+    val submissions = submissionService.findProjectKeyToSubmissionByReleaseName(releaseName);
+    submissionFileSystem.createInitialReleaseFilesystem(nextRelease, submissions, projects);
+  }
+
+  public boolean isSignOffAllowed(String releaseName) {
+    return submissionService.findSubmissionStatesByReleaseName(releaseName).stream()
+        .map(Submission::getState)
+        // At least one submission must be signed off on
+        .anyMatch(state -> state == SIGNED_OFF);
```
Comment (btiernay) [inline dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/service/ReleaseService.java]: j.u.f.Predicate#isEqual:`isEqual(SIGNED_OFF)`

File: dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/service/ReleaseService.java [hunk 4]
```diff
@@ -507,9 +457,14 @@ public void queueSubmissions(@NonNull List<QueuedProject> queuedProjects) throws
     release.enqueue(queuedProjects);
 
     val filePatternToTypeMap = dictionaryRepository.getFilePatternToTypeMap(release.getDictionaryVersion());
+    val projectKeys = queuedProjects.stream()
```
Comment (btiernay) [inline dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/service/ReleaseService.java:462]: Duplicated below. Lets add utility method.

File: dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/service/ReleaseService.java [hunk 5]
```diff
@@ -613,18 +577,25 @@ public void removeQueuedSubmissions(@NonNull String... targets) throws InvalidSt
...
+    val projectKeys = submissionService.findReleaseProjectKeys(getNextRelease().getName());
+    resetSubmissions(projectKeys);
   }
 
   @Synchronized
   public void resetInvalidSubmissions() {
     val release = getNextRelease();
     val filePatternToTypeMap = dictionaryRepository.getFilePatternToTypeMap(release.getDictionaryVersion());
-    for (val projectKey : release.getInvalidProjectKeys()) {
+    val invalidProjectKeys = submissionService.findSubmissionsByReleaseName(release.getName()).stream()
```
Comment (btiernay) [inline dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/service/ReleaseService.java]: This is very expensive (submissions are big!). Lets push down the predicate into the query.

File: dcc-submission-server/src/test/java/org/icgc/dcc/submission/server/service/ReleaseServiceTest.java [hunk 6]
```diff
@@ -59,12 +58,12 @@
 import com.mongodb.MongoClient;
 import com.mongodb.MongoException;
 
-import org.junit.Assert;
-import lombok.val;
-
 @RunWith(MockitoJUnitRunner.class)
+@Ignore("Add tests")
```
Comment (btiernay) [inline dcc-submission-server/src/test/java/org/icgc/dcc/submission/server/service/ReleaseServiceTest.java:62]: Lets remove this and make sure tests still pass.

**Review by btiernay — CHANGES_REQUESTED** [review]

File: dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/service/ReleaseService.java (hunk 4 above)
Comment (slovit) [reply dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/service/ReleaseService.java:462]: It's a 'getter'. It has no logic.

File: dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/service/ReleaseService.java (hunk 3 above)
Comment (slovit) [reply dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/service/ReleaseService.java]: What `Predicates`? There's no such a method neither in our commons `Predicates` nor in Guava's.

File: dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/repository/SubmissionRepository.java (hunk 2 above)
Comment (slovit) [reply dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/repository/SubmissionRepository.java:123]: I think 'no'. I didn't find how Morphia DSL could be used for DML. It looks like the version we use supports querying only.

**Review by btiernay — APPROVED** [review]

File: dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/service/ReleaseService.java [hunk 7]
```diff
@@ -613,7 +613,7 @@ public void resetInvalidSubmissions() {
     val release = getNextRelease();
     val filePatternToTypeMap = dictionaryRepository.getFilePatternToTypeMap(release.getDictionaryVersion());
     val invalidProjectKeys = submissionService.findSubmissionSummariesByReleaseName(release.getName()).stream()
-        .filter(isEqual(INVALID))
+        .filter(submission -> submission.getState() == INVALID)
```
Comment (btiernay) [inline dcc-submission-server/src/main/java/org/icgc/dcc/submission/server/service/ReleaseService.java:616]: Ah, my bad, sorry for leading you astray here!

File: dcc-submission-core/src/main/java/org/icgc/dcc/submission/core/security/Authorizations.java [hunk 8]
```diff
@@ -73,12 +87,33 @@ public static boolean hasSpecificProjectPrivilege(Authentication authentication,
...
+    if (authorities.contains(ALL_PROJECTS)) {
+      return ALL_PROJECTS_LIST;
+    }
+
+    return authorities.stream()
+        .map(Authorizations::extractProject)
+        .collect(toImmutableList());
+  }
+
+  private static String extractProject(String permission) {
```
Comment (btiernay) [inline dcc-submission-core/src/main/java/org/icgc/dcc/submission/core/security/Authorizations.java]: `parseProject`

**Pull request merged**